internal/controllers/core/extensionrepo/reconciler.go

```diff
@@ -100,59 +103,70 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+	var result ctrl.Result
 	if strings.HasPrefix(repo.Spec.URL, "file://") {
-		return r.reconcileFileRepo(ctx, &repo, strings.TrimPrefix(repo.Spec.URL, "file://"))
+		r.reconcileFileRepo(ctx, state, strings.TrimPrefix(repo.Spec.URL, "file://"))
+	} else {
+		// Check that the URL is valid.
+		importPath, err := getDownloaderImportPath(&repo)
+		if err != nil {
+			state.status = v1alpha1.ExtensionRepoStatus{Error: fmt.Sprintf("invalid: %v", err)}
+		} else {
+			result, err = r.reconcileDownloaderRepo(ctx, state, importPath)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
 	}
 
-	// Check that the URL is valid.
-	importPath, err := getDownloaderImportPath(&repo)
+	err = r.maybeUpdateStatus(ctx, &repo, state)
 	if err != nil {
-		return r.updateError(ctx, &repo, fmt.Sprintf("invalid: %v", err))
+		return ctrl.Result{}, err
 	}
-
-	return r.reconcileDownloaderRepo(ctx, nn, &repo, importPath, state)
+	return result, nil
 }
 
 // Reconcile a repo that lives on disk, and shouldn't otherwise be modified.
-func (r *Reconciler) reconcileFileRepo(ctx context.Context, repo *v1alpha1.ExtensionRepo, filePath string) (reconcile.Result, error) {
-	if repo.Spec.Ref != "" {
-		return r.updateError(ctx, repo, "spec.ref not supported on file:// repos")
+func (r *Reconciler) reconcileFileRepo(ctx context.Context, state *repoState, filePath string) {
+	if state.spec.Ref != "" {
+		state.status = v1alpha1.ExtensionRepoStatus{Error: "spec.ref not supported on file:// repos"}
+		return
 	}
 
 	if !filepath.IsAbs(filePath) {
-		msg := fmt.Sprintf("file paths must be absolute. Url: %s", repo.Spec.URL)
-		return r.updateError(ctx, repo, msg)
+		state.status = v1alpha1.ExtensionRepoStatus{
+			Error: fmt.Sprintf("file paths must be absolute. Url: %s", state.spec.URL),
+		}
+		return
 	}
 
 	info, err := os.Stat(filePath)
 	if err != nil {
-		msg := fmt.Sprintf("loading: %v", err)
-		return r.updateError(ctx, repo, msg)
+		state.status = v1alpha1.ExtensionRepoStatus{Error: fmt.Sprintf("loading: %v", err)}
+		return
 	}
 
 	if !info.IsDir() {
-		msg := "loading: not a directory"
-		return r.updateError(ctx, repo, msg)
+		state.status = v1alpha1.ExtensionRepoStatus{Error: "loading: not a directory"}
+		return
 	}
 
 	timeFetched := apis.NewTime(info.ModTime())
-	return r.updateStatus(ctx, repo, func(status *v1alpha1.ExtensionRepoStatus) {
-		status.Error = ""
-		status.LastFetchedAt = timeFetched
-		status.Path = filePath
-	})
+	state.status = v1alpha1.ExtensionRepoStatus{LastFetchedAt: timeFetched, Path: filePath}
 }
 
 // Reconcile a repo that we need to fetch remotely, and store
 // under ~/.tilt-dev.
-func (r *Reconciler) reconcileDownloaderRepo(ctx context.Context, nn types.NamespacedName,
-	repo *v1alpha1.ExtensionRepo, importPath string, state *repoState) (reconcile.Result, error) {
+func (r *Reconciler) reconcileDownloaderRepo(ctx context.Context, state *repoState, importPath string) (reconcile.Result, error) {
```

Comment:
It feels weird that this still returns `error` in a couple places (not sure if coincidental that they're both `stat` calls) instead of updating `state.status` accordingly

Reply:
i think you're right! i think this should be a status update instead of a retry